pyramid/view.py

```diff
@@ -414,3 +417,93 @@ def callback(context, name, ob):
         settings['_info'] = info.codeinfo # fbo "action_method"
         return wrapped
 
+def _find_views(
+    registry,
+    request_iface,
+    context_iface,
+    view_name,
+    view_types=None,
+    view_classifier=None,
+    ):
+    if  view_types is None:
+        view_types = (IView, ISecuredView, IMultiView)
+    if view_classifier is  None:
+        view_classifier = IViewClassifier
+    registered = registry.adapters.registered
+    cache = registry._view_lookup_cache
+    views = cache.get((request_iface, context_iface, view_name))
+    if views is None:
+        views = []
+        for req_type, ctx_type in itertools.product(
+            request_iface.__sro__, context_iface.__sro__
+        ):
+            source_ifaces = (view_classifier, req_type, ctx_type)
+            for view_type in view_types:
+                view_callable = registered(
+                    source_ifaces,
+                    view_type,
+                    name=view_name,
+                )
+                if view_callable is not None:
+                    views.append(view_callable)
+        if views:
+            # do not cache view lookup misses.  rationale: dont allow cache to
+            # grow without bound if somebody tries to hit the site with many
+            # missing URLs.  we could use an LRU cache instead, but then
+            # purposeful misses by an attacker would just blow out the cache
+            # anyway. downside: misses will almost always consume more CPU than
+            # hits in steady state.
+            with registry._lock:
+                cache[(request_iface, context_iface, view_name)] = views
+
+    return views
+
+def _call_view(
+    registry,
+    request,
+    context,
+    context_iface,
+    view_name,
+    view_types=None,
+    view_classifier=None,
+    secure=True,
+    request_iface=None,
+    ):
+    if request_iface is None:
+        request_iface = getattr(request, 'request_iface', IRequest)
+    view_callables = _find_views(
+        registry,
+        request_iface,
+        context_iface,
+        view_name,
+        view_types=view_types,
+        view_classifier=view_classifier,
+        )
+
+    pme = None
+    response = None
+
+    for view_callable in view_callables:
+        # look for views that meet the predicate criteria
+        try:
+            if not secure:
+                # the view will have a __call_permissive__ attribute if it's
+                # secured; otherwise it won't.
+                view_callable = getattr(
+                    view_callable,
+                    '__call_permissive__',
+                    view_callable
+                    )
+
+            # if this view is secured, it will raise a Forbidden
+            # appropriately if the executing user does not have the proper
+            # permission
+            response = view_callable(context, request)
+            return response
+        except PredicateMismatch as _pme:
```

Comment:
consider some way of returning all predicate mismatches, but as a separate issue from this pr